Add openclaw how-to (#267)
canonical/inference-snaps: docs/how-to/integration/openclaw.md

    webui: http://127.0.0.1:8339/
model:
    name: qwen3-8b-q4-k-m
```

Note down the `openai` endpoint URL and the `model.name` value, as you'll need them for configuration.

## Configure using the wizard

After installation, OpenClaw launches the interactive setup wizard with a number of sections to set up the agent.
The connection with an inference snap needs to be set up during the model configuration section.

If OpenClaw has already been set up, the wizard can be manually started again to run through the model configuration section:

```shell
openclaw configure --section model
```

The wizard will ask a number of questions.
Answer the questions using this guide as a reference:

- **Model/{spellexception}`auth` provider**: Choose "Custom Provider"
- **API Base URL**: Enter the `openai` endpoint URL obtained previously from the `status` command.
- **How do you want to provide this API key?**: Choose "Paste API key now"
- **API Key**: Leave blank
- **Endpoint compatibility**: Choose "OpenAI-compatible".
- **Model ID**: Enter the `model.name` value obtained previously from the `status` command.

The endpoint will now be verified by the wizard.

> If it fails, but the `<inference-snap> chat` command works, it is probably because the model does not have enough resources to be used by an agent.

Reviewer: ```
the model does not have enough resources to be used by an agent
```

This part is not entirely clear to me: does it mean that my computer does not have enough resources to run the chosen model as an agent (thus I should downgrade), or is the chosen model too "weak" to run agentic workflows (thus I should upgrade)?

Also, could you provide an example of failure? What does such a failure look like?

Author: Very good question. I was being vague on purpose. I've seen the check failing immediately on some models, if they take too long to respond. But on other models this check works, but later when you try and use it, they run out of memory and close the connection without responding (intel-gpu / ovms).

We could rephrase to be specific about the first issue, but I was trying to make a general statement to cover the second issue too.

- **Endpoint ID and Model alias**: You can give the endpoint and model alias any value of your choosing. We recommend making it clear this is the model from the inference snap.

The wizard should print `Configure complete`.
You can now use OpenClaw, and it should use the inference snap as its model.

If you have multiple models configured, the OpenClaw dashboard lets you choose the one to use from the dropdown below the chat input box.

## Configure using the configuration file

> OpenClaw's configuration is persisted to a file. One can also manually edit this file.

Reviewer: I would better point out that this section is an alternative workflow with respect to the previous section, something like the following:

```suggestion
An alternative way of configuring OpenClaw is by means of editing its configuration file.
```

(openclaw)=

# Use inference snaps in OpenClaw

Inference snaps that support tool calling can be used with local agents like OpenClaw, {spellexception}`PicoClaw`, {spellexception}`NemoClaw`, Hermes, etc.
The steps below are a guideline on how to configure OpenClaw to use an inference snap.
Similar steps can be used to configure other AI agents.

## Look up base URL and model ID

Run the respective inference snap's status command to look up the `openai` endpoint URL and the model name.
These values will be required when configuring the agent.

```shell

> $ <inference-snap> status

Author: Switched to terminal

For example, run:

```shell
<inference-snap> status

> ```

Reviewer 2: This guides the reader to expect an example, but instead there is a placeholder. The outputs below and others in this guide show the concrete example.

````suggestion
For example, using the `qwen3` snap:

```shell
qwen3 status
```
````

    openai: http://127.0.0.1:8338/v1
    webui: http://127.0.0.1:8339/
model:
    name: qwen3-8b-q4-k-m
```

## Configure using wizard

After installation, OpenClaw launches the interactive setup wizard with a number of sections to set up the agent.
The connection with an inference snap needs to be set up during the model configuration section.

If OpenClaw has already been set up, the wizard can be manually started again to run through the model configuration section:

```shell
openclaw configure --section model
```

The wizard will ask a number of questions.
Answer the questions using this guide as a reference.

Model/{spellexception}`auth` provider
~ Choose "Custom Provider"

API Base URL
~ Enter the `openai` endpoint URL obtained previously from the `status` command.

How do you want to provide this API key?
~ Choose "Paste API key now"

API Key
~ Leave blank

Endpoint compatibility
~ Choose either the legacy "OpenAI-compatible" or the newer "OpenAI Responses-compatible".

Model ID

> ~ Enter the model name obtained previously from the `status` command.

Author: It renders correctly as a description list

The wizard will now run a live completion probe to verify the endpoint.
If it fails, it could be due to insufficient RAM or VRAM, or missing tool-calling support in the model.
Refer to the [OpenClaw troubleshooting guide](https://docs.openclaw.ai/help/troubleshooting) for further help.

> - **Endpoint ID and Model alias**: You can give the endpoint and model alias any value of your choosing. We recommend making it clear this is the model from the inference snap.

Reviewer 2: Use direct instructions.
```suggestion
- **Endpoint ID and Model alias**: Set desired values, for example qwen3-snap and qwen3-8b, respectively.
```

I also suggest splitting them. The bullets above each map to 1 configuration field, and the combination here, in bold, makes it look like there is a single configuration step to fill both.

The wizard should print `Configure complete`.
You can now use OpenClaw, and it should use the inference snap as model.

If you have multiple models configured, the OpenClaw dashboard lets you choose the one to use from the dropdown below the chat input box.

## Configure using file

OpenClaw's configuration is persisted to a file. One can also manually edit this file.

```
nano ~/.openclaw/openclaw.json
```

After the wizard was run, the file will contain a section listing the available models.
You can manually edit this file to change the configuration or add new models.

```
{
...
  "models": {
    "mode": "merge",

>     "providers": {

Author: Added json and removed ellipses

> ```

Reviewer 2: Simplify wording. Drop the explicit editor (or make it clear this is an example).
```suggestion
An alternative way of configuring OpenClaw via its configuration file, typically found at `~/.openclaw/openclaw.json`. 
```

> After the wizard was run, the file will contain a section listing the available models.

Reviewer 2: It is a bit hard to follow. After reading this, I thought that file editing results in running the wizard.

Consider: "Add the inference snap as a model provider in the configuration file. If you have already run the wizard, the file will contain the models object. Otherwise, create the models object and add the inference snap as a provider."

Author: I decided to remove the entire section about manually editing the config file. It is not as simple as adding a model provider. You still need to add its ID to a list of models at the top of the file. If the wizard does not work, a user should refer to the OpenClaw docs.

>     "mode": "merge",

Author: This should not be copied and pasted directly, as it is just a snippet from the entire config file.

      "qwen3-snap": {
        "baseUrl": "http://127.0.0.1:8338/v1",
        "api": "openai-responses",
        "models": [
          {
            "id": "qwen3-8b-q4-k-m",
            "name": "qwen3-8b-q4-k-m (Custom Provider)",
            "input": [
              "text"
            ]
          }
        ]
      },
      "gemma4-snap": {

>         "baseUrl": "http://127.0.0.1:8336/v3",

Author: Depending on the selected engine, /v3 could be a valid path

        "api": "openai-completions",
        "models": [
          {
            "id": "gemma4-e4b-it-int4-ov",
            "name": "gemma4-e4b-it-int4-ov (Custom Provider)",
            "input": [
              "text",
              "image"
            ]
          }
        ]
      }
    }
  },
  // other configuration sections...
}

> ```

Reviewer 2: The example is too long IMO: Gemma4 is a distraction. I'd keep qwen3 with data matching what was being suggested in the wizard section for the endpoint id and model name.

Highlight affected lines with [emphasize-lines](https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html#parameterizing-directives-options), example:
https://github.com/canonical/inference-snaps/blob/c5edbc83155e56c63438b621f9d0c678227d9fdc/docs/how-to/integration/opencode.md?plain=1#L53-L56

If OpenClaw has already been set up, the wizard can be manually started again to run through the model configuration section:

```shell
openclaw configure --section model
```

The wizard will ask a number of questions.
Answer the questions using this guide as a reference:

- **Model/{spellexception}`auth` provider**: Under the "More…" section, choose "Custom Provider".
- **API Base URL**: Enter the `openai` endpoint URL obtained previously from the `status` command.
- **How do you want to provide this API key?**: Choose "Paste API key now"
- **API Key**: Leave blank
- **Endpoint compatibility**: Choose "OpenAI-compatible".
- **Model ID**: Enter the `model.name` value obtained previously from the `status` command.

The wizard will now run a live completion probe to verify the endpoint.
If it fails, it could be due to insufficient RAM or VRAM, or missing tool-calling support in the model.
Refer to the [OpenClaw troubleshooting guide](https://docs.openclaw.ai/help/troubleshooting) for further help.

- **Endpoint ID**: Any desired value, but we recommend making it clear this is the connection with the inference snap, e.g. `qwen3-snap-endpoint`.

> - **Model alias (optional)**: This is the name of the model that will be shown in the OpenClaw dashboard. If not set here, it will fall back to the model ID. We recommend making it clear this is the model from the inference snap, e.g. `qwen3-snap-model`.

Reviewer 2: As [mentioned](https://github.com/canonical/inference-snaps/pull/267#discussion_r3569201775), prefer direct instructions. In a how-to, the user should not be burdened with understanding and deciding on a naming convention. This is a multi-step, tedious wizard at steps 6-7. Because the wizard does not allow modifications to the entered fields (unless it fails), any decision-making can be taken seriously by the reader.

Provide a best practice that can be learned and applied elsewhere.

```suggestion
- **Endpoint ID**: Arbitrary identifier of the provider, e.g. set it to `qwen3-snap`.
- **Model alias (optional)**: An alias for the model. It falls back to the model ID if left unset. 
```
